snap/squashfs/stat.go

```diff
@@ -0,0 +1,330 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package squashfs
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+type SnapFileOwner struct {
+	UID uint32
+	GID uint32
+}
+
+type stat struct {
+	path  string
+	size  int64
+	mode  os.FileMode
+	mtime time.Time
+	user  string
+	group string
+}
+
+func (s stat) Name() string       { return filepath.Base(s.path) }
+func (s stat) Size() int64        { return s.size }
+func (s stat) Mode() os.FileMode  { return s.mode }
+func (s stat) ModTime() time.Time { return s.mtime }
+func (s stat) IsDir() bool        { return s.mode.IsDir() }
+func (s stat) Sys() interface{}   { return nil }
+func (s stat) Path() string       { return s.path } // not path of os.FileInfo
+
+const minLen = 57 // "drwxrwxr-x user/user             53595 2017-12-08 11:19 ."
+
+func fromRaw(raw []byte) (*stat, error) {
+	if len(raw) < minLen {
+		return nil, errBadLine(raw)
+	}
+
+	st := &stat{}
+
+	// first, the file mode, e.g. "-rwxr-xr-x"; always that length (1+3*3==10)
+	if err := st.parseMode(raw[:10]); err != nil {
+		return nil, err
+	}
+
+	// next, user/group info
+	p := 10
+	if n, err := st.parseOwner(raw[p:]); err != nil {
+		return nil, err
+	} else {
+		p += n
+	}
+
+	// next'll come the size or the node type
+	if n, err := st.parseSize(raw[p:]); err != nil {
+		return nil, err
+	} else {
+		p += n
+	}
+
+	if err := st.parseTime(raw[p : p+16]); err != nil {
+		return nil, err
+	}
+
+	p += 16
+
+	if raw[p] != ' ' {
```

Comment:
I wonder if this and the following lines should be st.parsePath()  - mostly for reasons to have stuff look similar?